backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java

```diff
@@ -28,16 +33,81 @@ public DefaultUserService(UserDao userDao) {
 
     @Override
     public User getUserById(Long id) {
+        LOG.info("Retrieved user ID {}", id);
         return userDao.findUserById(id);
     }
 
+    /**
+     * Find a unique User with email and username
+     * @param email The email of the user
+     * @param username The username of the user
+     * @return The User in the database
+     */
+    @Override
+    public User getUserByEmailAndUsername(String email, String username) {
+        List<User> userList = userDao.findAll();
+        User userToFind = null;
+
+        for (User user : userList) {
+            if (email.equalsIgnoreCase(user.getEmail()) && username.equalsIgnoreCase(user.getUsername())) {
+                userToFind = user;
+                LOG.info("Found user with email {}", user.getEmail());
+                break;
+            }
+        }
+
+        return userToFind;
+    }
+
     @Override
     public User createUser(User user) {
+        String password = user.getPassword();
+        String encryptedPassword = encryptData(password);
+        user.setPassword(encryptedPassword);
+        userDao.save(user);
+        LOG.info("Created user: {}", user.getFirstName());
+
+        return user;
+    }
 
-        User createdUser = userDao.save(user);
+    /**
+     * Method to look if user exists with unique email and username.
+     * @param email The email coming from the frontend.
+     * @param username The username coming from the frontend.
+     * @return boolean if exists or not.
+     */
+    @Override
+    public boolean userExists(String email, String username) {
+        boolean exists = false;
+        User user = userDao.findUserByEmailAndUsername(email, username);
+
+        if (user != null) {
+            exists = true;
+            LOG.info("User already exists.");
+        }
+        return exists;
+    }
+
+    @Override
+    public User registerAlpacaApiKey(User userWithKey) {
+        String key = userWithKey.getAlpacaApiKey();
+        String email = userWithKey.getEmail();
+        String username = userWithKey.getUsername();
+        User userToUpdate = userDao.findUserByEmailAndUsername(email, username);
+        if (userToUpdate != null) {
+            userToUpdate.setAlpacaApiKey(key);
+            userDao.save(userToUpdate);
+            LOG.info("Registered Alpaca key for user email {}", userToUpdate.getEmail());
+        }
+
+        return userToUpdate;
+    }
 
-        LOG.info("Created user: {}", createdUser.getFirstName());
+    public String encryptData(String text) {
+        return secureString.toGraphProperty(text);
+    }
 
-        return createdUser;
+    public String decryptData(String text) {
+        return secureString.toEntityAttribute(text);
```

Comment:
Do we need this? We shouldn't have to decrypt password and manipulate the unencrypted string, it is bad practice.

Reply:
True. I thought initially that when we login in the future we would compare the decrypted passwords but it is not recommended to do that and we should compare the digests instead.